### PR TITLE
apps: introduce App Engine interface

### PR DESCRIPTION
**Pull request by mike-sul**
Introduce App Engine interface which simplify testing of app related
updates as well as allows encapsulating of other engine
types/variations, e.g. ostree-based apps.

Signed-off-by: Mike Sul <mike.sul@foundries.io>

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -1,8 +1,26 @@
 set(TARGET_EXE aktualizr-lite)
 set(TARGET ${MAIN_TARGET_LIB})
 
-set(SRC helpers.cc liteclient.cc composeappmanager.cc docker.cc composeapp.cc composeapptree.cc ostree/sysroot.cc ostree/repo.cc)
-set(HEADERS helpers.h liteclient.h composeappmanager.h docker.h composeapp.h composeapptree.h ostree/sysroot.h ostree/repo.h)
+set(SRC helpers.cc
+        composeappengine.cc
+        appengine.cc
```
Comment (doanac) [inline src/CMakeLists.txt:6]: this is an empty file. is a place holder or something ?

File: src/composeappmanager.cc [hunk 2]
```diff
@@ -68,11 +66,13 @@ ComposeAppManager::ComposeAppManager(const PackageConfig& pconfig, const Bootloa
...
 
   const auto& current_apps = getApps(getCurrent());
+  // TODO: refactor it to fix "app rollback when ostree rollback" issue
   for (const auto& app_pair : current_apps) {
     const auto& app_name = app_pair.first;
-    auto need_start_flag = cfg_.apps_root / app_name / Docker::ComposeApp::NeedStartFile;
+    auto need_start_flag = cfg_.apps_root / app_name / Docker::ComposeAppEngine::NeedStartFile;
     if (boost::filesystem::exists(need_start_flag)) {
-      app_ctor_(app_name).start();
+      // app_ctor_(app_name).start();
```
Comment (doanac) [inline src/composeappmanager.cc:74]: can delete that line now

File: src/appengine.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    std::string name;
+    std::string uri;
+  };
+
+ public:
+  virtual bool fetch(const App& app) = 0;
+  virtual bool install(const App& app) = 0;
+  virtual bool run(const App& app) = 0;
+  virtual void remove(const App& app) = 0;
+  virtual bool isInstalled(const App& app) = 0;
```
Comment (doanac) [inline src/appengine.h:18]: This doesn't seem to be used *yet*. I say drop it until you introduce code that will use it.

**Review by doanac — COMMENTED** [review]
Looks pretty good. A couple of minor things.

Also - test broke with:
~~~
101: C++ exception with description "boost::filesystem::canonical: No such file or directory: "/usr/bin/docker"" thrown in the test body.
101: [  FAILED  ] helpers.lite_client_finalize (42 ms)
101: [ RUN      ] helpers.target_has_tags
101: [       OK ] helpers.target_has_tags (6 ms)
101: [ RUN      ] helpers.locking
~~~

File: src/appengine.h (hunk 3 above)
Comment (mike-sul) [reply src/appengine.h:18]: It's used in composeappengine.cc/h, The methods are defined as 'override' there. If I remove them the code won't compile.

File: src/composeappmanager.cc (hunk 2 above)
Comment (mike-sul) [reply src/composeappmanager.cc:74]: indeed.

**Comment by mike-sul** [comment]
> Also - test broke with:

It passes locally, has to do with the test container env specifics, will fix it anyway.

File: src/CMakeLists.txt (hunk 1 above)
Comment (mike-sul) [reply src/CMakeLists.txt:6]: yes, I can remove it.

File: src/appengine.h (hunk 3 above)
Comment (doanac) [reply src/appengine.h:18]: but its dead code there also, right? i see it *defined* and has a TODO, but its not actually referenced by any code yet?

Comment (mike-sul) [reply src/appengine.h:18]: no, it's not, except isInstalled().

**Comment by mike-sul** [comment]
@doanac This is initial set of tests I wanted to introduce, so that's all changes in the scope of the given PR.

File: src/composeappmanager.h [hunk 4]
```diff
@@ -62,10 +60,10 @@ class ComposeAppManager : public OstreeManager {
  private:
   Config cfg_;
   std::shared_ptr<OSTree::Sysroot> sysroot_;
-  Docker::RegistryClient registry_client_;
+  // Docker::RegistryClient registry_client_;
```
Comment (doanac) [inline src/composeappmanager.h:63]: remove that now.

File: tests/test_liteclient.cc [hunk 5]
```diff
@@ -32,6 +37,24 @@ static std::string run_cmd(const std::string &executable_to_run,
...
+      ON_CALL(*this, run).WillByDefault(Return(true));
+      ON_CALL(*this, isRunning).WillByDefault(Return(true));
+    }
+  }
+ public:
+  MOCK_METHOD(bool, fetch, (const App& app), (override));
+  MOCK_METHOD(bool, install, (const App& app), (override));
+  MOCK_METHOD(bool, run, (const App& app), (override));
+  MOCK_METHOD(void, remove, (const App& app), (override));
+  MOCK_METHOD(bool, isRunning, (const App& app), (const, override));
```
Comment (doanac) [inline tests/test_liteclient.cc:55]: mind blown

**Review by doanac — COMMENTED** [review]
just one silly thing. the mocking stuff is really cool. nice. 

feel free to fix and merge.

**Pull request merged**